### ISSUE TITLE
Roku client appears to NOT update NFO file after video finishes playing through to the very end

### ISSUE THREAD

**Issue by thecodygriffin** [jellyfin-archive/jellyfin-roku-legacy#1814]
## Software Versions
- Jellyfin Server Version:  10.9.7
- Roku Client Version:  2.0.7

## Describe the bug
Upon allowing a video to play through to the very end, the Roku client appears to NOT prompt the update of the corresponding NFO file.

To be very specific, "watched" is NOT updating from "false" to "true" in the corresponding NFO file.

## Prerequisites To Reproduce
1. Navigate to the Libraries > Libraries section of the JellyFin server from the Dashboard in the main navigation
2. Select the "Manage library" option in the kebab menu for the testing library
3. Select the "Nfo" checkbox in the "Metadata savers" section for the testing library
4. Select "OK" to save the setting
5. Navigate to the Libraries > NFO Settings section of the JellyFin server in the main navigation menu
6. Select the the user to be used for testing in the "Save user watch data to NFO files for" field
7. Select "Save" to save the setting

## How To Reproduce
1. Open the JellyFin Roku client
2. Ensure you are logged in with the same user selected in Step 6 of the "Prerequisites To Reproduce" section
3. Navigate to the library selected in Step 2 of the "Prerequisites To Reproduce" section
4. Select any video that has not yet been watched
5. Watch the video through to the very end when the stream closes and returns you to the library

## Expected behavior
The NFO file is updated including "playcount" and "watched" among others when applicable

## Logs
There are no [ERR] messages in the logs for my testing time range. There was an instance of the following [WRN] message, but it does not appear to correlate with the behavior experienced.

`[2024-07-03 10:52:25.892 -04:00] [WRN] WS "10.16.8.1" error receiving data: "The remote party closed the WebSocket connection without completing the close handshake."`

## Connection Information
- Is server local or remote? **LOCAL**
- Is server connection HTTP or HTTPS? **HTTP**

## Additional context
If the video is exited by the user prior to conclusion, the NFO file is updated as expected whether that be the "watched", "playcount","lastplayed", or "resume" values.

Furthermore, the behavior is NOT reproducible from the Android or Web clients.

**Comment by roidy** [comment]
Maybe related to the following issue:- #1567 

The app handles the `stopped` state which happens when the user ends a file by pressing back on the remote but fails to handle the `finished` state which is fired when the media naturally stops at the end with no user interaction.

**Comment by cewert** [comment]
@thecodygriffin can you give https://github.com/jellyfin/jellyfin-roku/pull/1941 a test when you get a chance?

**Comment by thecodygriffin** [comment]
@cewert Is it fair to assume that I will need to build the Roku app with the #1941 pull request and then side load it to test?

If so, I may not be able to do that myself for a few weeks. Otherwise, if there is an existing test build you can point me to, I can side load it sometime this weekend and test.

**Comment by cewert** [comment]
> @cewert Is it fair to assume that I will need to build the Roku app with the #1941 pull request and then side load it to test?

Yes, you will need to sideload the app to test but you don't have to build it unless you want to use the debugger and see debugging output. Our CI automatically creates a build zip for all PRs https://github.com/jellyfin/jellyfin-roku/actions/runs/11086513250?pr=1941

**Comment by roidy** [comment]
> Our CI automatically creates a build zip for all PRs https://github.com/jellyfin/jellyfin-roku/actions/runs/11086513250?pr=1941

How exactly do I build that? It's missing a ton of files.... No make file, no config files, no package file :(

**Comment by roidy** [comment]
Unfortunately that doesn't fix my problem of the video player not returning to SDR mode after playing a HDR/DV file to completion. In fact it's still not registering a finished state.

**Comment by roidy** [comment]
Here's a video showing the issue:- https://streamable.com/d4wqgk

Watching the activity in the Jellyfin server you can see that when stopping media using the back button on the remote the server registers a finished state, however when the media naturally stops at the end with no user interaction no finished state is being sent to the server and the server still thinks the video is active.

**Comment by thecodygriffin** [comment]
@cewert similar to @roidy findings, the build from https://github.com/jellyfin/jellyfin-roku/actions/runs/11086513250?pr=1941 did not resolve this issue. The NFO file is still not updating upon the "natural" conclusion of a video.

**Comment by cewert** [comment]
@thecodygriffin There was a bug in that build. Sorry for not giving you the new link. Can you try this [build](https://github.com/jellyfin/jellyfin-roku/actions/runs/11087871463?pr=1941) as well? That's the current build for #1941

**Comment by thecodygriffin** [comment]
@cewert Success! The latest build from https://github.com/jellyfin/jellyfin-roku/actions/runs/11087871463?pr=1941 resolves the issue.

I confirmed that the NFO file was updated in the following scenarios:
- Upon a video "naturally" concluding and returning me back to the library
- Upon a video "naturally" concluding and auto playing the next video in a season

I also verified that the NFO file is still updated when existing it early prior to the conclusion.

Thank you for your work on this. I can't wait until the fix makes it into an upcoming release.
